File: python/meromorphic/transfer_fn.py
"""NumPy mirror of ``src/lib/transferFn.ts``.

Modal-coordinate SISO transfer functions: ``H(z) = sum_k r_k / (z - p_k)``.
Poles/residues are complex128 arrays. Polynomial work uses :func:`numpy.poly`
and :func:`numpy.roots` for numerical stability; the explicit Durand-Kerner
mirror lives in :mod:`meromorphic.poly_roots` for parity testing.
"""

from __future__ import annotations

import math

import numpy as np

from .poly_roots import organize_roots


def _as_complex(arr) -> np.ndarray:
    return np.asarray(arr, dtype=np.complex128)


def eval_h(poles: np.ndarray, residues: np.ndarray, z: complex) -> complex:
    poles = _as_complex(poles)
    residues = _as_complex(residues)
    if poles.size == 0:
        return complex(0.0, 0.0)
    return complex(np.sum(residues / (z - poles)))


def frequency_response(
    poles: np.ndarray, residues: np.ndarray, n: int
) -> tuple[np.ndarray, np.ndarray]:
    theta = 2 * np.pi * np.arange(n) / n
    z = np.exp(1j * theta)
    poles = _as_complex(poles)
    residues = _as_complex(residues)
    # broadcast: (n, 1) - (k,) -> (n, k)
    h = np.sum(residues[None, :] / (z[:, None] - poles[None, :]), axis=1)
    return theta, h


def numerator_poly(poles: np.ndarray, residues: np.ndarray) -> np.ndarray:
    """Return Q(z) coefficients in ascending order, matching the JS layout.

    JS ``numeratorPoly`` returns ``polyAdd``-accumulated terms each of length n
    (since ``fromRoots(others)`` over n-1 roots gives degree n-1, length n).
    To match length-for-length we pad the accumulator to length n even if the
    leading coefficient happens to cancel.
    """
    poles = _as_complex(poles)
    residues = _as_complex(residues)
    n = poles.size
    if n == 0:
        return np.zeros(1, dtype=np.complex128)
    acc = np.zeros(n, dtype=np.complex128)
    for k in range(n):
        others = np.delete(poles, k)
        # np.poly returns descending coefficients of prod (z - others).
        term_desc = np.poly(others) if others.size > 0 else np.array([1.0 + 0j])
        term_asc = term_desc[::-1].astype(np.complex128) * residues[k]
        acc[: term_asc.size] += term_asc
    return acc


def denominator_poly(poles: np.ndarray) -> np.ndarray:
    poles = _as_complex(poles)
    if poles.size == 0:
        return np.array([1.0 + 0j], dtype=np.complex128)
    desc = np.poly(poles)
    return desc[::-1].astype(np.complex128)


def zeros(poles: np.ndarray, residues: np.ndarray) -> np.ndarray:
    q = numerator_poly(poles, residues)
    if q.size <= 1:
        return np.zeros(0, dtype=np.complex128)
    deg = q.size - 1
    while deg > 0 and abs(q[deg]) < 1e-12:
        deg -= 1
    if deg == 0:
        return np.zeros(0, dtype=np.complex128)
    trimmed_desc = q[: deg + 1][::-1]
    rts = np.roots(trimmed_desc)
    return organize_roots(rts)


def impulse_response(poles: np.ndarray, residues: np.ndarray, t: int) -> np.ndarray:
    poles = _as_complex(poles)
    residues = _as_complex(residues)
    g = np.zeros(t, dtype=np.float64)
    for k in range(poles.size):
        powk = complex(1.0, 0.0)
        pk = complex(poles[k])
        rk = complex(residues[k])
        for s in range(t):
            g[s] += (rk * powk).real
            powk = powk * pk
    return g


def effective_rank(residues: np.ndarray, threshold: float = 1e-6) -> int:
    residues = _as_complex(residues)
    return int(np.sum(np.abs(residues) > threshold))


def pole_zero_separation(poles: np.ndarray, residues: np.ndarray) -> np.ndarray:
    poles = _as_complex(poles)
    zs = zeros(poles, residues)
    out = np.full(poles.size, np.inf, dtype=np.float64)
    if zs.size == 0:
        return out
    for k in range(poles.size):
        d = np.abs(poles[k] - zs)
        out[k] = float(np.min(d))
    return out


def simulate(poles: np.ndarray, residues: np.ndarray, x: np.ndarray) -> np.ndarray:
    poles = _as_complex(poles)
    residues = _as_complex(residues)
    x = np.asarray(x, dtype=np.float64)
    t = x.size
    n = poles.size
    y = np.zeros(t, dtype=np.float64)
    phi = np.zeros(n, dtype=np.complex128)
    for i in range(t):
        phi = poles * phi + complex(x[i], 0.0)
        y[i] = float(np.sum(residues * phi).real)
    return y

File: src/lib/transferFn.ts
/*
 * Transfer function utilities for SISO systems.
 *
 * A SISO transfer function with simple poles is fully specified by:
 *   - poles p_1, ..., p_n in C (closed under complex conjugation for real systems)
 *   - residues r_1, ..., r_n in C (with r_k conjugate-paired alongside p_k)
 *
 * Then H(z) = sum_k r_k / (z - p_k), and the numerator polynomial is
 *   Q(z) = sum_k r_k * prod_{j != k} (z - p_j),
 * whose roots are the zeros of the system. The full impulse response is
 *   g_s = sum_k r_k p_k^s.
 *
 * Working in the (poles, residues) modal parameterisation lets us avoid
 * eigendecomposition entirely for the static demos (sections 4-6). Section 7's
 * training loop will swap this layer for an explicit (W, b, c) gradient flow.
 */

import {
  type Complex,
  ZERO,
  ONE,
  add,
  mul,
  sub,
  div,
  scale,
  abs,
  expi,
  polyVal,
  polyMul,
  polyAdd,
  polyScale,
  fromRoots,
} from './complex';
import { polyRoots, organizeRoots } from './polyRoots';

export interface ModalSystem {
  poles: Complex[];      // length n
  residues: Complex[];   // length n
}

/** Evaluate H(z) at a single complex z. */
export function evalH(sys: ModalSystem, z: Complex): Complex {
  let acc: Complex = { ...ZERO };
  for (let k = 0; k < sys.poles.length; k++) {
    const denom = sub(z, sys.poles[k]);
    acc = add(acc, div(sys.residues[k], denom));
  }
  return acc;
}

/** Evaluate H on the unit circle at N equally-spaced angles in [0, 2pi). */
export function frequencyResponse(sys: ModalSystem, N: number): { theta: number[]; H: Complex[] } {
  const theta: number[] = new Array(N);
  const H: Complex[] = new Array(N);
  for (let i = 0; i < N; i++) {
    const th = (2 * Math.PI * i) / N;
    theta[i] = th;
    H[i] = evalH(sys, expi(th));
  }
  return { theta, H };
}

/** Build the numerator polynomial Q(z) = sum_k r_k * prod_{j != k} (z - p_j),
 *  in ascending coefficient form. Returns degree n-1 polynomial (or lower if
 *  cancellations occur). */
export function numeratorPoly(sys: ModalSystem): Complex[] {
  const n = sys.poles.length;
  if (n === 0) return [ZERO];

  let acc: Complex[] = [];
  for (let k = 0; k < n; k++) {
    // Build prod_{j != k} (z - p_j)
    const others = sys.poles.filter((_, j) => j !== k);
    const term = polyScale(fromRoots(others), sys.residues[k]);
    acc = acc.length === 0 ? term : polyAdd(acc, term);
  }
  return acc;
}

/** Denominator polynomial P(z) = prod_k (z - p_k). */
export function denominatorPoly(sys: ModalSystem): Complex[] {
  return fromRoots(sys.poles);
}

/** Zeros of H(z): roots of the numerator polynomial. */
export function zeros(sys: ModalSystem): Complex[] {
  const q = numeratorPoly(sys);
  if (q.length <= 1) return [];
  // Trim near-zero leading coefficients to avoid spurious "roots at infinity".
  let deg = q.length - 1;
  while (deg > 0 && abs(q[deg]) < 1e-12) deg--;
  if (deg === 0) return [];
  const trimmed = q.slice(0, deg + 1);
  return organizeRoots(polyRoots(trimmed));
}

/** Impulse response g_s for s = 0, ..., T-1. */
export function impulseResponse(sys: ModalSystem, T: number): number[] {
  // g_s = sum_k r_k p_k^s. The imaginary parts cancel for real systems.
  const g = new Array<number>(T).fill(0);
  for (let k = 0; k < sys.poles.length; k++) {
    let powk: Complex = { ...ONE };
    for (let s = 0; s < T; s++) {
      const term = mul(sys.residues[k], powk);
      g[s] += term.re;
      powk = mul(powk, sys.poles[k]);
    }
  }
  return g;
}

/** Effective rank: number of modes with |residue| above threshold. */
export function effectiveRank(sys: ModalSystem, threshold = 1e-6): number {
  let count = 0;
  for (const r of sys.residues) {
    if (abs(r) > threshold) count++;
  }
  return count;
}

/**
 * Pole-zero separation for the k-th pole: distance to its nearest zero in C.
 * Used for the saddle-distance / effective-residue analyses in §5–§7.
 */
export function poleZeroSeparation(sys: ModalSystem): number[] {
  const zs = zeros(sys);
  const out = new Array(sys.poles.length).fill(Infinity);
  for (let k = 0; k < sys.poles.length; k++) {
    for (const z of zs) {
      const d = abs(sub(sys.poles[k], z));
      if (d < out[k]) out[k] = d;
    }
  }
  return out;
}

/**
 * Apply an arbitrary input sequence through the system and return the output.
 * Uses modal decomposition: y_t = sum_k r_k * phi_k(t), where
 * phi_k(t) = sum_{s=0..t-1} p_k^s x_{t-s}.
 */
export function simulate(sys: ModalSystem, x: readonly number[]): number[] {
  const T = x.length;
  const n = sys.poles.length;
  const y = new Array<number>(T).fill(0);
  // Carry phi_k(t) as a running complex state for each mode.
  const phi: Complex[] = new Array(n);
  for (let k = 0; k < n; k++) phi[k] = { ...ZERO };
  for (let t = 0; t < T; t++) {
    // phi_k <- p_k * phi_k + x_t
    for (let k = 0; k < n; k++) {
      const next = add(mul(sys.poles[k], phi[k]), { re: x[t], im: 0 });
      phi[k] = next;
    }
    // y_t = sum_k r_k * phi_k  (imag parts cancel for real systems)
    let acc = 0;
    for (let k = 0; k < n; k++) {
      const term = mul(sys.residues[k], phi[k]);
      acc += term.re;
    }
    y[t] = acc;
  }
  return y;
}
